### PR TITLE
provides a optional-plugin for tracking ignores enhancement(#1203)

### PR DESCRIPTION
**Pull request by SevenPointOld**
Please answer these questions before submitting pull request

- Why submit this pull request?

New feature provided

- Related issues
#1203 
___

___
### New feature or improvement
- The applications can set paths in the plugin cofig file to let agent ignore tracing, use ant path match 
- This plugin after packing will create its own directory in optional-plugins parent directory, this directory has README.md file(introduce how to use this plugin), config file(plugin config) and plugin jar. I  also write an unit test for the plugin

### CODE REVIEW

**Comment by liuhaoyang** [comment]
Maybe we can also support reading `trace.ignore_path` from environment variables.

File: apm-sniffer/optional-plugins/pom.xml [hunk 1]
```diff
@@ -98,11 +100,26 @@
                         <configuration>
                             <tasks>
                                 <taskdef resource="net/sf/antcontrib/antcontrib.properties" classpathref="maven.runtime.classpath" />
+                                <mkdir dir="${plugin.dest.dir}" />
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/pom.xml]: Please read the pom.xml in `apm-sdk-plugin` module and plugins in it. Create pom properties `agent.package.dest.dir` and `optional.agent.package.dest.dir` to do the copy, and let copy op could be override by the plugin pom.

Only copy jar file is required, the setting file and readme are optional.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+###How to use trace ignore plugin
+Please just copy or move the apm-trace-ignore-plugin-x.jar to agent/plugins, then you can set you need ignore paths in trace-ignore-plugin.config(this config file don't need to move)
```
[the remark this answers is not in the document]
Comment (wu-sheng) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/README.md]: Setting file should be in `/conf` folder.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/TraceIgnoreExtendService.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.skywalking.apm.plugin.trace.ignore.conf.IgnoreConfig;
+import org.apache.skywalking.apm.plugin.trace.ignore.conf.IgnoreConfigInitializer;
+import org.apache.skywalking.apm.plugin.trace.ignore.matcher.SpringAntPathMatcher;
+import org.apache.skywalking.apm.plugin.trace.ignore.matcher.TracePathMatcher;
+import org.apache.skywalking.apm.util.StringUtil;
+
+/**
+ *
+ * @author liujc [liujunc1993@163.com]
+ * @date 2018/5/11 11:27
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/TraceIgnoreExtendService.java]: Don't put the date in Java file. No one will change it.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/TracePathMatcher.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package org.apache.skywalking.apm.plugin.trace.ignore.matcher;
+
+public interface TracePathMatcher {
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/TracePathMatcher.java]: I can't find this matcher used in any place. What is the purpose of this interface?

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package org.apache.skywalking.apm.plugin.trace.ignore.matcher;
+
+import org.springframework.util.AntPathMatcher;
+import org.springframework.util.PathMatcher;
+
+public class SpringAntPathMatcher implements TracePathMatcher {
+
+    private PathMatcher pathMatcher = new AntPathMatcher();
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java]: If you are using the Spring matcher, then the plugin has to include the Spring jar into it. And you put `<artifactId>spring-core</artifactId>` provided in your pom. So who can make the spring core jar in the agent class loader?

Comment (wu-sheng) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java]: The further question is, is using spring core matcher a good idea? Include such a large library to solve this feature.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/README.md]: I agree that.

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/TracePathMatcher.java (hunk 4 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/TracePathMatcher.java]: I think there may be a variety of path matching rules. This interface is designed to support this extension

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java (hunk 5 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java]: I've had this thought too, I will not depend on the spring framework

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/TraceIgnoreExtendService.java (hunk 3 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/TraceIgnoreExtendService.java]: Ok, I'll remove the date

File: apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java (hunk 5 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/matcher/SpringAntPathMatcher.java]: I will add a simple version of AntPathMatcher

File: apm-sniffer/optional-plugins/pom.xml (hunk 1 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/pom.xml]: I get it

**Comment by SevenPointOld** [comment]
I have revised code and pushed to repository, please review the new request

File: apm-sniffer/optional-plugins/trace-ignore-plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+###How to use trace ignore plugin
+Please copy the apm-trace-ignore-plugin-x.jar to `agent/plugins`, this plugin support reading config from environment variables(The env key must start with `skywalking.`, the reuslt should be as same as in `apm-trace-ignore-plugin.config`), or you can copy the `apm-trace-ignore-plugin.config` to `agent/config` then you'll set you need ignore paths in `apm-trace-ignore-plugin.config`
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/README.md]: See here: https://github.com/SevenPointOld/incubator-skywalking/blob/18790363b3ba50738ea3e577b63506fce16a4e77/apm-sniffer/optional-plugins/trace-ignore-plugin/README.md

Your markdown don't look good. Please provide sections about two usages.
```
# title
Abstrct

## Usage
1. ...... use config file.....
2. ...... through env, and examples
```

File: apm-sniffer/optional-plugins/pom.xml [hunk 7]
```diff
@@ -31,11 +31,15 @@
         <shade.net.bytebuddy.source>net.bytebuddy</shade.net.bytebuddy.source>
         <shade.net.bytebuddy.target>${shade.package}.${shade.net.bytebuddy.source}</shade.net.bytebuddy.target>
 
-        <plugin.dest.dir>${project.build.directory}/../../../../skywalking-agent/optional-plugins</plugin.dest.dir>
+        <sdk.plugin.related.dir />
+        <agent.package.dest.dir>${project.build.directory}${sdk.plugin.related.dir}/../../../../skywalking-agent</agent.package.dest.dir>
+        <optional.agent.package.dest.dir>${agent.package.dest.dir}/optional-plugins</optional.agent.package.dest.dir>
+        <directory.plugins>apm-trace-ignore-plugin</directory.plugins>
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/pom.xml]: `directory.plugins` should not in root pom.xml

File: apm-sniffer/optional-plugins/pom.xml [hunk 8]
```diff
@@ -98,11 +102,22 @@
...
                             <tasks>
                                 <taskdef resource="net/sf/antcontrib/antcontrib.properties" classpathref="maven.runtime.classpath" />
+                                <mkdir dir="${optional.agent.package.dest.dir}" />
                                 <if>
                                     <equals arg1="${project.packaging}" arg2="jar" />
                                     <then>
-                                        <mkdir dir="${plugin.dest.dir}" />
-                                        <copy file="${project.build.directory}/${project.artifactId}-${project.version}.jar" tofile="${plugin.dest.dir}/${project.artifactId}-${project.version}.jar" overwrite="true" />
+                                        <if>
+                                            <!-- if the plugin is a directory -->
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/pom.xml]: I am wondering, whether be better if put conf and jar in the `optional-plugins` dictionary directly? I feel strange about this part. Or do you have any suggestion?

File: apm-sniffer/optional-plugins/trace-ignore-plugin/pom.xml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                                <groupId>ant</groupId>
+                                <artifactId>ant</artifactId>
+                            </exclusion>
+                        </exclusions>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.apache.ant</groupId>
+                        <artifactId>ant-nodeps</artifactId>
+                        <version>1.8.1</version>
+                    </dependency>
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/trace-ignore-plugin/pom.xml]: Would these two will in your optional jar? Could you provide a screenshot to show me what are in it?

Comment (wu-sheng) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/pom.xml]: If not, there should be a problem.

If so, you need to learn the policies about redistribution jars in Apache project.

File: apm-sniffer/optional-plugins/pom.xml [hunk 10]
```diff
@@ -31,11 +31,15 @@
         <shade.net.bytebuddy.source>net.bytebuddy</shade.net.bytebuddy.source>
         <shade.net.bytebuddy.target>${shade.package}.${shade.net.bytebuddy.source}</shade.net.bytebuddy.target>
 
-        <plugin.dest.dir>${project.build.directory}/../../../../skywalking-agent/optional-plugins</plugin.dest.dir>
+        <sdk.plugin.related.dir />
```
[the remark this answers is not in the document]
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/pom.xml]: I add  `<sdk.plugin.related.dir />` follow `apm-sdk-plugin/pom`  
image
It's just a statement, the sub pom if need this properties should override it,  then act on `${agent.package.dest.dir}`

File: apm-sniffer/optional-plugins/trace-ignore-plugin/README.md (hunk 6 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/README.md]: ok

File: apm-sniffer/optional-plugins/pom.xml (hunk 8 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/pom.xml]: My mistake, i'll remove that

File: apm-sniffer/optional-plugins/trace-ignore-plugin/pom.xml (hunk 9 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/trace-ignore-plugin/pom.xml]: My mistake, I'll fix that

File: apm-sniffer/optional-plugins/pom.xml (hunk 8 above)
Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/pom.xml]: Well, I think the meaning of this plugin may be somewhat confusing,
One is to integrate other components to achieve link tracking.
The other is to extend SPI's own behavior.

Comment (SevenPointOld) [reply apm-sniffer/optional-plugins/pom.xml]: If there is a standard example of an extension to SPI, it may be clear.

Comment (wu-sheng) [reply apm-sniffer/optional-plugins/pom.xml]: I am following. I just have concerns about this pom. Too many your plugin details expose to root pom.

**Comment by SevenPointOld** [comment]
This is current `optional-plugins` look like
image
In the `AntPathMatcher` , the most of the code is copied form Spring Framework. I only remove the superfluous method, and make some revise  in order to pass through  `checkstyle`.
I've never had such experience before, and I don't know what the effect of adding Apache license to the code will be.
I'm sorry about that. If this can not be merged, I think it might be possible to use these extended functions in the form of third party plugins.

**Comment by wu-sheng** [comment]
> I'm sorry about that. If this can not be merged, I think it might be possible to use these extended functions in the form of third party plugins.

@SevenPointOld There is no need to sorry. But we do need to deal with source codes used, rather then distribution(*.jar) used in this cases. 

I am going to ask our Apache mentor @michaelsembwever  to some suggestions. But I believe, we have to adjust LICENSE and NOTICE files in root.

**Comment by wu-sheng** [comment]
> This is current optional-plugins look like

Sorry, I didn't make myself clear. I mean what look like in distribution? After your package, you should find it in `dist` folder, and unzip it.

**Comment by SevenPointOld** [comment]
> After your package, you should find it in dist folder, and unzip it.

After unzip it,  the whole distribution like this
image

**Comment by wu-sheng** [comment]
@SevenPointOld Good to share. I think it is good. The last thing before merge, is we have to process the LICENSE in right way. Let's wait our mentor's suggestion.

**Comment by michaelsembwever** [comment]
@wu-sheng, i do not know how strict ASF would be on this, because Spring is also under Apache 2.0 license. But you can't just transfer the copyright like that.

So to be safe I would take the approach of using `AntPathMatcher` from the original dependency. That original dependency can be stripped down to just the one class, and the package name refactored for safety. This can be done using either, or both, the maven dependency and shade plugins.

**Comment by wu-sheng** [comment]
@SevenPointOld please use shade plugin to transfer this class from spring denpendency, rather then use code copy. like this https://maven.apache.org/plugins/maven-shade-plugin/examples/class-relocation.html Please notice, only transfer the only one you need.

Based on this, I can trust on License.

**Comment by SevenPointOld** [comment]
@wu-sheng  Should I create a new module as a dependent transformation?

**Comment by wu-sheng** [comment]
@SevenPointOld I think should add in this plugin module.

**Comment by SevenPointOld** [comment]
> I think should add in this plugin module.

What about the distribution look like this?
image

**Comment by wu-sheng** [comment]
Why do you want the redistribition in the separated module and jar?

**Comment by SevenPointOld** [comment]
> Please notice, only transfer the only one you need
It's hard to transfer the only one, because of the Spring `AntPathMatcher` import some util class, the other util class also import others,  but they all belong to a same package, can I transfer the whole package `org.springframework.util`? exclude child package of course

**Comment by wu-sheng** [comment]
@SevenPointOld Could you finish this PR soon? I look forward to merge this pr into beta2 release.

**Comment by SevenPointOld** [comment]
@wu-sheng  I'll try to rewrite the match logic.

**Comment by wu-sheng** [comment]
>  I'll try to rewrite the match logic.

Try to make it a simple and easy one, at the first edition.

**Comment by SevenPointOld** [comment]
Promise

**Comment by wu-sheng** [comment]
Merged, hope you could provide a blog to tell people to use it.

**Comment by wu-sheng** [comment]
@SevenPointOld I think you miss add your README to incubating features menu.

image

Also, you should consider to add a section in Chinese document: https://github.com/apache/incubator-skywalking/blob/master/docs/README_ZH.md  The title should be `孵化特性`.

**Comment by SevenPointOld** [comment]
OK,  I'll do it

**Pull request merged**